### PR TITLE
Separate Depthwidth and Lossguide growing policy in fast histogram

### PR DESCRIPTION
**Pull request by CodingCat**
it's a following PR of https://github.com/dmlc/xgboost/pull/4011 which separates depthwidth and loss guide growing policy in fast histogram per our discussion in https://github.com/dmlc/xgboost/issues/4077

closes https://github.com/dmlc/xgboost/issues/4077

@trivialfis @hcho3 @RAMitchell please help to review

### CODE REVIEW

File: jvm-packages/dev/build.sh [hunk 1]
```diff
@@ -17,5 +17,5 @@ rm /usr/bin/python
 ln -s /opt/rh/python27/root/usr/bin/python /usr/bin/python
 
 # build xgboost
-cd /xgboost/jvm-packages;mvn package
+cd /xgboost/jvm-packages;ulimit -c unlimited;mvn package
```
Comment (hcho3) [inline jvm-packages/dev/build.sh]: Why `ulimit -c unlimited` here? Did you run into stack space issue?

Comment (CodingCat) [reply jvm-packages/dev/build.sh]: yes, it is a script used in docker container, I found (at least in my environment) that the stacktrace cannot be shown when debugging issues like code crashed in native layer

**Comment by hcho3** [comment]
@CodingCat Can you rebase against the latest master?

**Comment by CodingCat** [comment]
@trivialfis @hcho3 @RAMitchell ping for review

**Comment by CodingCat** [comment]
@trivialfis @hcho3 @RAMitchell ping for review

**Comment by hcho3** [comment]
Will review today

File: src/common/column_matrix.h [hunk 2]
```diff
@@ -71,7 +71,6 @@ class ColumnMatrix {
                 double  sparse_threshold) {
     const auto nfeature = static_cast<bst_uint>(gmat.cut.row_ptr.size() - 1);
     const size_t nrow = gmat.row_ptr.size() - 1;
-
```
Comment (RAMitchell) [inline src/common/column_matrix.h]: There seems to be a few unnecessary formatting changes in this PR. It makes it harder to review as we have to search through these carefully to see if anything is actually changed, as well as adding unnecessary history to these files.

File: src/tree/updater_quantile_hist.cc [hunk 3]
```diff
@@ -89,39 +88,163 @@ bool QuantileHistMaker::UpdatePredictionCache(
...
   unsigned timestamp = 0;
+  int num_leaves = 0;
 
-  double tstart;
-  double time_init_data = 0;
-  double time_init_new_node = 0;
-  double time_build_hist = 0;
-  double time_evaluate_split = 0;
-  double time_apply_split = 0;
+  tstart = dmlc::GetTime();
```
Comment (RAMitchell) [inline src/tree/updater_quantile_hist.cc]: I would like to see these timers replaced with the Monitor class.

**Comment by trivialfis** [comment]
Glanced the changes.  Will leave comments today.  Got into some troubles with my network access recently.  Sorry for the delay.

File: src/tree/updater_quantile_hist.cc [hunk 4]
```diff
@@ -89,39 +88,163 @@ bool QuantileHistMaker::UpdatePredictionCache(
...
+    "tree_method=hist does not support multiple roots at this moment";
 
+  // in depth_wise growing, we feed loss_chg with 0.0 since it is not used anyway
+  qexpand_depth_wise.push_back(ExpandEntry(0, p_tree->GetDepth(0), 0.0, timestamp++));
+  ++num_leaves;
+  // get stats of node per depth
   tstart = dmlc::GetTime();
-  this->InitData(gmat, gpair_h, *p_fmat, *p_tree);
-  time_init_data = dmlc::GetTime() - tstart;
+  for (int depth = 0; depth < param_.max_depth + 1; depth++) {
```
Comment (RAMitchell) [inline src/tree/updater_quantile_hist.cc]: The logic inside this for loop contains many different cases and will be difficult to read and maintain. Can we further simplify this? If I understand correctly most of this code is evaluating whether to actually build the histogram or defer construction of the histogram and use the subtraction trick.

File: src/tree/updater_quantile_hist.cc [hunk 5]
```diff
@@ -89,39 +88,163 @@ bool QuantileHistMaker::UpdatePredictionCache(
...
+        } else if (IsRoot(nid)) {
+          // root node
+          hist_.AddHistRow(nid);
+          BuildHist(gpair_h, row_set_collection_[nid], gmat, gmatb, hist_[nid], false);
+          sync_count++;
+          starting_index = std::min(starting_index, nid);
+        }
+      }
+    }
+    // 1. sync histogram
```
Comment (RAMitchell) [inline src/tree/updater_quantile_hist.cc]: If you have to number your comments to describe the flow of this function perhaps these numbered sections should be functions on their own.

File: src/tree/updater_quantile_hist.h [hunk 6]
```diff
@@ -97,13 +97,16 @@ class QuantileHistMaker: public TreeUpdater {
...
-                          GHistRow hist) {
+                          GHistRow hist,
+                          bool sync_hist) {
       if (param_.enable_feature_grouping > 0) {
         hist_builder_.BuildBlockHist(gpair, row_indices, gmatb, hist);
       } else {
         hist_builder_.BuildHist(gpair, row_indices, gmat, hist);
       }
-      this->histred_.Allreduce(hist.data(), hist_builder_.GetNumBins());
+      if (sync_hist) {
```
Comment (RAMitchell) [inline src/tree/updater_quantile_hist.h]: Is there some penalty for calling this unnecessarily?

File: src/tree/updater_quantile_hist.h [hunk 7]
```diff
@@ -218,13 +235,41 @@ class QuantileHistMaker: public TreeUpdater {
     const DMatrix* p_last_fmat_;
 
     using ExpandQueue =
-        std::priority_queue<ExpandEntry, std::vector<ExpandEntry>,
-                            std::function<bool(ExpandEntry, ExpandEntry)>>;
-    std::unique_ptr<ExpandQueue> qexpand_;
+       std::priority_queue<ExpandEntry, std::vector<ExpandEntry>,
+                           std::function<bool(ExpandEntry, ExpandEntry)>>;
+
+    std::unique_ptr<ExpandQueue> qexpand_loss_guided;
```
Comment (hcho3) [inline src/tree/updater_quantile_hist.h]: I think we should suffix all class variables with `_` to differentiate them from local variables. Example: `qexpand_loss_guided_`.

File: src/tree/updater_quantile_hist.cc [hunk 8]
```diff
@@ -89,39 +88,163 @@ bool QuantileHistMaker::UpdatePredictionCache(
...
+    std::vector<ExpandEntry> temp_qexpand_depth;
+    // 0. build histogram for the nodes in current level
+    for (size_t k = 0; k < qexpand_depth_wise.size(); k++) {
+      int nid = qexpand_depth_wise[k].nid;
+      if (rabit::IsDistributed()) {
+        if (IsLeft(nid) || IsRoot(nid)) {
+          // in distributed setting, we always calcuate from left child or root node
+          hist_.AddHistRow(nid);
+          BuildHist(gpair_h, row_set_collection_[nid], gmat, gmatb, hist_[nid], false);
+          if (left_to_right_siblings.find(nid) != left_to_right_siblings.end()) {
```
Comment (hcho3) [inline src/tree/updater_quantile_hist.cc]: Use `IsLeft()` here?

File: src/tree/updater_quantile_hist.cc [hunk 9]
```diff
@@ -89,39 +88,163 @@ bool QuantileHistMaker::UpdatePredictionCache(
...
+        hist_[(*p_tree)[local_it->first].Parent()]);
+    }
+    time_build_hist += dmlc::GetTime() - tstart;
+    // 2. initNewNode to get it's stats
+    tstart = dmlc::GetTime();
+    for (size_t k = 0; k < qexpand_depth_wise.size(); k++) {
+      int nid = qexpand_depth_wise[k].nid;
+      this->InitNewNode(nid, gmat, gpair_h, *p_fmat, *p_tree);
+      // add constraints
+      if (right_to_left_siblings.find(nid) != right_to_left_siblings.end()) {
```
Comment (hcho3) [inline src/tree/updater_quantile_hist.cc]: Use `IsRight()` here?

**Comment by CodingCat** [comment]
@RAMitchell @hcho3  thank you very much for the review, I have addressed most of the comments (will continue working on the performance monitor class tmr )

File: jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostGeneralSuite.scala [hunk 10]
```diff
@@ -111,7 +111,31 @@ class XGBoostGeneralSuite extends FunSuite with PerTest {
     assert(eval.eval(model._booster.predict(testDM, outPutMargin = true), testDM) < 0.1)
   }
 
-  test("test with fast histo with monotone_constraints") {
+  test("test with fast histo with monotone_constraints (lossguide)") {
```
Comment (trivialfis) [inline jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostGeneralSuite.scala]: `fast hist` is renamed as `quantile hist`.

File: jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostGeneralSuite.scala [hunk 11]
```diff
@@ -111,7 +111,31 @@ class XGBoostGeneralSuite extends FunSuite with PerTest {
...
+    val eval = new EvalError()
+    val training = buildDataFrame(Classification.train)
+    val testDM = new DMatrix(Classification.test.iterator)
+    val paramMap = Map("eta" -> "1",
+      "max_depth" -> "6", "silent" -> "1",
```
Comment (trivialfis) [inline jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostGeneralSuite.scala]: `silent` is deprecated, use `verbosity` instead.

File: src/common/hist_util.cc [hunk 12]
```diff
@@ -1,8 +1,6 @@
 /*!
- * Copyright 2017-2018 by Contributors
+ * Copyright 2019 by Contributors
```
Comment (trivialfis) [inline src/common/hist_util.cc]: 2017~2019?

File: src/tree/updater_quantile_hist.cc [hunk 13]
```diff
@@ -89,39 +88,195 @@ bool QuantileHistMaker::UpdatePredictionCache(
...
-  int num_leaves = 0;
-  unsigned timestamp = 0;
+void QuantileHistMaker::Builder::BuildLocalHistograms(
+    int &starting_index,
+    int &sync_count,
+    const GHistIndexMatrix &gmat,
+    const GHistIndexBlockMatrix &gmatb,
+    RegTree *p_tree,
+    const std::vector<GradientPair> &gpair_h) {
+  // 0. build histogram for the nodes in current level
```
Comment (trivialfis) [inline src/tree/updater_quantile_hist.cc]: If you want to outline the steps, please gather the descriptions into one place.  The next item is 500 lines away...

File: src/tree/updater_quantile_hist.cc [hunk 14]
```diff
@@ -89,39 +88,195 @@ bool QuantileHistMaker::UpdatePredictionCache(
...
+          nodes_to_derive_[(*p_tree)[node.Parent()].RightChild()] = nid;
+        }
+        sync_count++;
+        starting_index = std::min(starting_index, nid);
+      }
+    } else {
+      if (!node.IsRoot() && node.IsLeftChild() &&
+          (row_set_collection_[nid].Size() <
+           row_set_collection_[(*p_tree)[node.Parent()].RightChild()].Size())) {
+        hist_.AddHistRow(nid);
```
Comment (trivialfis) [inline src/tree/updater_quantile_hist.cc]: Please move
``` c++
        hist_.AddHistRow(nid);
        BuildHist(gpair_h, row_set_collection_[nid], gmat, gmatb, hist_[nid], false);
```
out of the testing branches, I think they are the same in every branch.

File: src/tree/updater_quantile_hist.cc [hunk 15]
```diff
@@ -89,39 +88,195 @@ bool QuantileHistMaker::UpdatePredictionCache(
...
+  // 0. build histogram for the nodes in current level
+  for (size_t k = 0; k < qexpand_depth_wise_.size(); k++) {
+    int nid = qexpand_depth_wise_[k].nid;
+    RegTree::Node &node = (*p_tree)[nid];
+    if (rabit::IsDistributed()) {
+      if (node.IsRoot() || node.IsLeftChild()) {
+        // in distributed setting, we always calcuate from left child or root node
+        hist_.AddHistRow(nid);
+        BuildHist(gpair_h, row_set_collection_[nid], gmat, gmatb, hist_[nid], false);
+        if (!node.IsRoot() && node.IsLeftChild()) {
```
Comment (trivialfis) [inline src/tree/updater_quantile_hist.cc]: `node.IsLeftChild()` is true.  No need to test.

File: src/tree/updater_quantile_hist.cc [hunk 16]
```diff
@@ -89,39 +88,195 @@ bool QuantileHistMaker::UpdatePredictionCache(
...
+  unsigned timestamp = 0;
+  int num_leaves = 0;
 
   tstart = dmlc::GetTime();
+
   this->InitData(gmat, gpair_h, *p_fmat, *p_tree);
   time_init_data = dmlc::GetTime() - tstart;
 
+  // FIXME(hcho3): this code is broken when param.num_roots > 1. Please fix it
+  CHECK_EQ(p_tree->param.num_roots, 1) <<
```
Comment (trivialfis) [inline src/tree/updater_quantile_hist.cc]: This check is duplicated many times.

File: src/tree/updater_quantile_hist.cc [hunk 17]
```diff
@@ -89,39 +88,195 @@ bool QuantileHistMaker::UpdatePredictionCache(
...
+  const ColumnMatrix &column_matrix,
+  DMatrix *p_fmat,
+  RegTree *p_tree,
+  const std::vector<GradientPair> &gpair_h) {
+  unsigned timestamp = 0;
+  int num_leaves = 0;
 
   tstart = dmlc::GetTime();
+
   this->InitData(gmat, gpair_h, *p_fmat, *p_tree);
```
Comment (trivialfis) [inline src/tree/updater_quantile_hist.cc]: This is called in `Update`.  It might be duplicated call, please verify.

File: src/tree/updater_quantile_hist.cc [hunk 18]
```diff
@@ -89,39 +88,195 @@ bool QuantileHistMaker::UpdatePredictionCache(
...
+    // 2. initNewNode to get it's stats
+    tstart = dmlc::GetTime();
+    BuildNodeStats(gmat, p_fmat, p_tree, gpair_h);
+    time_init_new_node += dmlc::GetTime() - tstart;
+    // 3. evaluateSplit
+    EvaluateSplits(gmat, column_matrix, p_fmat, p_tree, num_leaves, depth, timestamp,
+            temp_qexpand_depth);
+    // 4. if qexpand is empty and temp_qexpand is empty, break
+    if (temp_qexpand_depth.empty()) {
+      qexpand_depth_wise_.clear();
```
Comment (trivialfis) [inline src/tree/updater_quantile_hist.cc]: Please move
``` c++
      qexpand_depth_wise_.clear();
      nodes_to_derive_.clear();
```
out of testing branches.

File: src/tree/updater_quantile_hist.cc [hunk 19]
```diff
@@ -183,25 +338,51 @@ void QuantileHistMaker::Builder::Update(const GHistIndexMatrix& gmat,
...
+  tstart = dmlc::GetTime();
+  this->InitData(gmat, gpair_h, *p_fmat, *p_tree);
+  time_init_data = dmlc::GetTime() - tstart;
+
+  // FIXME(hcho3): this code is broken when param.num_roots > 1. Please fix it
+  CHECK_EQ(p_tree->param.num_roots, 1)
+      << "tree_method=hist does not support multiple roots at this moment";
+  if (param_.grow_policy == TrainParam::kLossGuide) {
+     ExpandWithLossGuide(gmat, gmatb, column_matrix, p_fmat, p_tree, gpair_h);
+    while (!qexpand_loss_guided_->empty()) {
```
Comment (trivialfis) [inline src/tree/updater_quantile_hist.cc]: Can we move this loop into `ExpandWithLossGuide`?

File: src/tree/updater_quantile_hist.cc [hunk 20]
```diff
@@ -183,25 +338,51 @@ void QuantileHistMaker::Builder::Update(const GHistIndexMatrix& gmat,
...
+      const int nid = qexpand_loss_guided_->top().nid;
+      qexpand_loss_guided_->pop();
+      (*p_tree)[nid].SetLeaf(snode_[nid].weight * param_.learning_rate);
+    }
+  } else {
+    ExpandWithDepthWidth(gmat, gmatb, column_matrix, p_fmat, p_tree, gpair_h);
+  }
 
   // set all the rest expanding nodes to leaf
```
Comment (trivialfis) [inline src/tree/updater_quantile_hist.cc]: Is this comment still valid?

File: src/tree/updater_quantile_hist.h [hunk 21]
```diff
@@ -127,18 +141,18 @@ class QuantileHistMaker: public TreeUpdater {
                        const RegTree& tree);
 
     void ApplySplit(int nid,
-                    const GHistIndexMatrix& gmat,
-                    const ColumnMatrix& column_matrix,
-                    const HistCollection& hist,
-                    const DMatrix& fmat,
-                    RegTree* p_tree);
+                     const GHistIndexMatrix& gmat,
```
Comment (trivialfis) [inline src/tree/updater_quantile_hist.h]: Please see these indents on Github, your editor might need some adjustments.

File: src/tree/updater_quantile_hist.h [hunk 22]
```diff
@@ -218,19 +255,33 @@ class QuantileHistMaker: public TreeUpdater {
...
     using ExpandQueue =
-        std::priority_queue<ExpandEntry, std::vector<ExpandEntry>,
-                            std::function<bool(ExpandEntry, ExpandEntry)>>;
-    std::unique_ptr<ExpandQueue> qexpand_;
+       std::priority_queue<ExpandEntry, std::vector<ExpandEntry>,
+                           std::function<bool(ExpandEntry, ExpandEntry)>>;
+
+    std::unique_ptr<ExpandQueue> qexpand_loss_guided_;
+    std::vector<ExpandEntry> qexpand_depth_wise_;
+    // key is the node id which should be calculated by SubstractTrick, value is the node is which
```
Comment (trivialfis) [inline src/tree/updater_quantile_hist.h]: > value is the node is which

Typo?

File: src/tree/updater_quantile_hist.h [hunk 23]
```diff
@@ -218,19 +255,33 @@ class QuantileHistMaker: public TreeUpdater {
...
+    // provides the evidence for substracts
+    std::unordered_map<int, int> nodes_to_derive_;
```
Comment (trivialfis) [inline src/tree/updater_quantile_hist.h]: I think there might be a better word than "derive".

File: src/tree/updater_quantile_hist.cc [hunk 24]
```diff
@@ -89,39 +88,195 @@ bool QuantileHistMaker::UpdatePredictionCache(
...
 
-void QuantileHistMaker::Builder::Update(const GHistIndexMatrix& gmat,
-                                        const GHistIndexBlockMatrix& gmatb,
-                                        const ColumnMatrix& column_matrix,
-                                        HostDeviceVector<GradientPair>* gpair,
-                                        DMatrix* p_fmat,
-                                        RegTree* p_tree) {
-  double gstart = dmlc::GetTime();
+void QuantileHistMaker::Builder::SyncHistograms(
+    int &starting_index,
```
Comment (trivialfis) [inline src/tree/updater_quantile_hist.cc]: I should probably get clang-tidy working as soon as possible, there is no need to pass as reference.

File: src/tree/updater_quantile_hist.cc [hunk 25]
```diff
@@ -89,39 +88,195 @@ bool QuantileHistMaker::UpdatePredictionCache(
...
+
+  unsigned timestamp = 0;
+  int num_leaves = 0;
+
   // FIXME(hcho3): this code is broken when param.num_roots > 1. Please fix it
   CHECK_EQ(p_tree->param.num_roots, 1)
-      << "tree_method=hist does not support multiple roots at this moment";
+    << "tree_method=hist does not support multiple roots at this moment";
+
   for (int nid = 0; nid < p_tree->param.num_roots; ++nid) {
```
Comment (trivialfis) [inline src/tree/updater_quantile_hist.cc]: So `num_roots` is partially supported?  @hcho3 Is there any plan to add full support?  Otherwise we might just remove it.

File: src/tree/updater_quantile_hist.cc [hunk 26]
```diff
@@ -183,25 +338,51 @@ void QuantileHistMaker::Builder::Update(const GHistIndexMatrix& gmat,
...
                                  snode_[cright].best.loss_chg,
                                  timestamp++));
 
       ++num_leaves;  // give two and take one, as parent is no longer a leaf
     }
   }
+}
+
+void QuantileHistMaker::Builder::Update(const GHistIndexMatrix& gmat,
+                                    const GHistIndexBlockMatrix& gmatb,
```
Comment (trivialfis) [inline src/tree/updater_quantile_hist.cc]: Is your editor mixing tab and space?

File: src/tree/updater_quantile_hist.cc (hunk 14 above)
[the remark this answers is not in the document]
Comment (CodingCat) [reply src/tree/updater_quantile_hist.cc]: actually no, for example 

if this is a left child and its number of samples is larger than its sibling's, we should do nothing and defer it to the time when we visit the sibling

if we move these lines out of test branches, it will be added in anyway

File: src/tree/updater_quantile_hist.cc (hunk 19 above)
[the remark this answers is not in the document]
Comment (CodingCat) [reply src/tree/updater_quantile_hist.cc]: I actually feel this loop is not necessary, @hcho3 ?

**Comment by trivialfis** [comment]
Seems the comments got a little bit messy.  It's because my machine got frozen during a review.  After rebooting, the previous review was not visible to me so I have to do it again.  After publishing it the previous somehow shows up ... If you similar comments, please ignore one of them.

File: src/tree/updater_quantile_hist.cc (hunk 3 above)
Comment (CodingCat) [reply src/tree/updater_quantile_hist.cc]: I have added the class

**Comment by CodingCat** [comment]
I think I have addressed all comments, ready for the next round of review

**Comment by RAMitchell** [comment]
Is there some reason you didn't use the existing Monitor class?
https://github.com/dmlc/xgboost/blob/99a290489cb8e7109ff4f8f1161f098185097b25/src/common/timer.h#L41

**Comment by CodingCat** [comment]
@RAMitchell  I was not aware of this class...but when I take a look at Monitor class there, I found it has several differences with the performance monitoring here, e.g. the format of the report and the log level, I am a bit hesitated about whether we should change the behavior of performance reporting either in hist or in other places using that class

I personally vote to unify the performance monitoring across updaters after this release

File: src/tree/updater_quantile_hist.cc (hunk 19 above)
[the remark this answers is not in the document]
Comment (hcho3) [reply src/tree/updater_quantile_hist.cc]: @CodingCat This loop exists so that the queue `qexpand_loss_guided_` gets emptied properly at the end. Note that some choice of split evaluators may result in early termination of the algorithm, with a bunch of leaf nodes left in the queue. This loop ensures that the nodes actually get labelled as leaf.

File: src/tree/updater_quantile_hist.cc (hunk 25 above)
[the remark this answers is not in the document]
Comment (hcho3) [reply src/tree/updater_quantile_hist.cc]: Opened #4133 to deprecate `num_roots` parameter.

File: src/tree/updater_quantile_hist.cc (hunk 19 above)
[the remark this answers is not in the document]
Comment (CodingCat) [reply src/tree/updater_quantile_hist.cc]: but for loss guide, we always jump out from the while loop https://github.com/dmlc/xgboost/blob/master/src/tree/updater_quantile_hist.cc#L140, do you mean in other future cases?

File: src/tree/updater_quantile_hist.cc (hunk 20 above)
Comment (hcho3) [reply src/tree/updater_quantile_hist.cc]: This comment should be moved to line 359

File: src/tree/updater_quantile_hist.cc [hunk 27]
```diff
@@ -89,119 +88,286 @@ bool QuantileHistMaker::UpdatePredictionCache(
...
+
+  perf_monitor.TickStart();
+  this->InitData(gmat, gpair_h, *p_fmat, *p_tree);
+  perf_monitor.UpdatePerfTimer(TreeGrowingPerfMonitor::timer_name::INIT_DATA);
+
+  // FIXME(hcho3): this code is broken when param.num_roots > 1. Please fix it
+  CHECK_EQ(p_tree->param.num_roots, 1)
+      << "tree_method=hist does not support multiple roots at this moment";
+  if (param_.grow_policy == TrainParam::kLossGuide) {
+     ExpandWithLossGuide(gmat, gmatb, column_matrix, p_fmat, p_tree, gpair_h);
```
Comment (hcho3) [inline src/tree/updater_quantile_hist.cc]: Fix the indentation here?

File: src/tree/updater_quantile_hist.cc (hunk 19 above)
[the remark this answers is not in the document]
Comment (hcho3) [reply src/tree/updater_quantile_hist.cc]: You're actually right; the loop is currently not needed. See comments 
https://github.com/dmlc/xgboost/blob/9b917cda4ff61f0512cdba1c53bc8e55258a5d4a/src/tree/updater_quantile_hist.cc#L197-L199

File: src/tree/updater_quantile_hist.h [hunk 28]
```diff
@@ -114,6 +119,94 @@ class QuantileHistMaker: public TreeUpdater {
                                HostDeviceVector<bst_float>* p_out_preds);
 
    protected:
+    /* tree growing policies */
+    struct ExpandEntry {
+        int nid;
```
Comment (hcho3) [inline src/tree/updater_quantile_hist.h]: Why 4 spaces indent here? We should stick to 2 spaces

File: src/tree/updater_quantile_hist.h [hunk 29]
```diff
@@ -114,6 +119,94 @@ class QuantileHistMaker: public TreeUpdater {
...
+        int depth;
+        bst_float loss_chg;
+        unsigned timestamp;
+        ExpandEntry(int nid, int depth, bst_float loss_chg, unsigned tstmp)
+                : nid(nid), depth(depth), loss_chg(loss_chg), timestamp(tstmp) {}
+    };
+
+    struct TreeGrowingPerfMonitor {
+        enum timer_name {INIT_DATA, INIT_NEW_NODE, BUILD_HIST, EVALUATE_SPLIT, APPLY_SPLIT};
```
Comment (hcho3) [inline src/tree/updater_quantile_hist.h]: 2 spaces instead of 4

File: src/tree/updater_quantile_hist.cc (hunk 19 above)
[the remark this answers is not in the document]
Comment (hcho3) [reply src/tree/updater_quantile_hist.cc]: Let's just remove this loop, along with the 3-line comments?

File: src/tree/updater_quantile_hist.cc (hunk 20 above)
Comment (hcho3) [reply src/tree/updater_quantile_hist.cc]: NVM, let's just remove the comment

File: src/tree/updater_quantile_hist.cc [hunk 30]
```diff
@@ -89,119 +88,286 @@ bool QuantileHistMaker::UpdatePredictionCache(
...
+
+  const std::vector<GradientPair>& gpair_h = gpair->ConstHostVector();
+
+  spliteval_->Reset();
+
+  perf_monitor.TickStart();
+  this->InitData(gmat, gpair_h, *p_fmat, *p_tree);
+  perf_monitor.UpdatePerfTimer(TreeGrowingPerfMonitor::timer_name::INIT_DATA);
+
+  // FIXME(hcho3): this code is broken when param.num_roots > 1. Please fix it
```
Comment (hcho3) [inline src/tree/updater_quantile_hist.cc]: Remove this, since we're deprecating `num_roots`

File: src/tree/updater_quantile_hist.h [hunk 31]
```diff
@@ -218,13 +334,20 @@ class QuantileHistMaker: public TreeUpdater {
...
-                            std::function<bool(ExpandEntry, ExpandEntry)>>;
-    std::unique_ptr<ExpandQueue> qexpand_;
+       std::priority_queue<ExpandEntry, std::vector<ExpandEntry>,
+                           std::function<bool(ExpandEntry, ExpandEntry)>>;
+
+    std::unique_ptr<ExpandQueue> qexpand_loss_guided_;
+    std::vector<ExpandEntry> qexpand_depth_wise_;
+    // key is the node id which should be calculated by SubstractTrick, value is the node which
+    // provides the evidence for substracts
+    std::unordered_map<int, int> nodes_for_substrack_trick;
```
Comment (hcho3) [inline src/tree/updater_quantile_hist.h]: Let's use correct spelling:
```suggestion
    std::unordered_map<int, int> nodes_for_subtraction_trick;
```

File: src/tree/updater_quantile_hist.cc [hunk 32]
```diff
@@ -89,119 +88,286 @@ bool QuantileHistMaker::UpdatePredictionCache(
...
+        (*starting_index) = std::min((*starting_index), nid);
+      }
+    } else {
+      if (!node.IsRoot() && node.IsLeftChild() &&
+          (row_set_collection_[nid].Size() <
+           row_set_collection_[(*p_tree)[node.Parent()].RightChild()].Size())) {
+        hist_.AddHistRow(nid);
+        BuildHist(gpair_h, row_set_collection_[nid], gmat, gmatb, hist_[nid], false);
+        nodes_for_substrack_trick[(*p_tree)[node.Parent()].RightChild()] = nid;
+        sync_count++;
```
Comment (hcho3) [inline src/tree/updater_quantile_hist.cc]: Did you mean to use `(*sync_count)++;` here?

File: src/tree/updater_quantile_hist.cc [hunk 33]
```diff
@@ -89,119 +88,286 @@ bool QuantileHistMaker::UpdatePredictionCache(
...
+        (*starting_index) = std::min((*starting_index), nid);
+      } else if (!node.IsRoot() && !node.IsLeftChild() &&
+                 (row_set_collection_[nid].Size() <=
+                  row_set_collection_[(*p_tree)[node.Parent()].LeftChild()].Size())) {
+        hist_.AddHistRow(nid);
+        BuildHist(gpair_h, row_set_collection_[nid], gmat, gmatb, hist_[nid], false);
+        nodes_for_substrack_trick[(*p_tree)[node.Parent()].LeftChild()] = nid;
+        sync_count++;
```
Comment (hcho3) [inline src/tree/updater_quantile_hist.cc]: Did you mean to use `(*sync_count)++;` here?

File: src/tree/updater_quantile_hist.cc [hunk 34]
```diff
@@ -89,119 +88,286 @@ bool QuantileHistMaker::UpdatePredictionCache(
...
+        (*starting_index) = std::min((*starting_index), nid);
+      } else if (node.IsRoot()) {
+        // root node
+        hist_.AddHistRow(nid);
+        BuildHist(gpair_h, row_set_collection_[nid], gmat, gmatb, hist_[nid], false);
+        sync_count++;
```
Comment (hcho3) [inline src/tree/updater_quantile_hist.cc]: Did you mean to use `(*sync_count)++;` here?

File: src/tree/updater_quantile_hist.cc [hunk 35]
```diff
@@ -89,119 +88,286 @@ bool QuantileHistMaker::UpdatePredictionCache(
...
-                                        DMatrix* p_fmat,
-                                        RegTree* p_tree) {
-  double gstart = dmlc::GetTime();
+void QuantileHistMaker::Builder::SyncHistograms(
+    int starting_index,
+    int sync_count,
+    RegTree *p_tree) {
+  perf_monitor.TickStart();
+  this->histred_.Allreduce(hist_[starting_index].data(), hist_builder_.GetNumBins() * sync_count);
+  // use substracttricks
```
Comment (hcho3) [inline src/tree/updater_quantile_hist.cc]:
```suggestion
  // use Subtraction Trick
```

File: src/tree/updater_quantile_hist.cc (hunk 27 above)
Comment (hcho3) [reply src/tree/updater_quantile_hist.cc]:
```suggestion
    ExpandWithLossGuide(gmat, gmatb, column_matrix, p_fmat, p_tree, gpair_h);
```

File: src/tree/updater_quantile_hist.h [hunk 36]
```diff
@@ -218,13 +334,20 @@ class QuantileHistMaker: public TreeUpdater {
...
     using ExpandQueue =
-        std::priority_queue<ExpandEntry, std::vector<ExpandEntry>,
-                            std::function<bool(ExpandEntry, ExpandEntry)>>;
-    std::unique_ptr<ExpandQueue> qexpand_;
+       std::priority_queue<ExpandEntry, std::vector<ExpandEntry>,
+                           std::function<bool(ExpandEntry, ExpandEntry)>>;
+
+    std::unique_ptr<ExpandQueue> qexpand_loss_guided_;
+    std::vector<ExpandEntry> qexpand_depth_wise_;
+    // key is the node id which should be calculated by SubstractTrick, value is the node which
```
Comment (hcho3) [inline src/tree/updater_quantile_hist.h]:
```suggestion
    // key is the node id which should be calculated by Subtraction Trick, value is the node which
```

File: src/tree/updater_quantile_hist.cc [hunk 37]
```diff
@@ -89,119 +88,286 @@ bool QuantileHistMaker::UpdatePredictionCache(
...
+        (param_.max_leaves > 0 && (*num_leaves) == param_.max_leaves)) {
+      (*p_tree)[nid].SetLeaf(snode_[nid].weight * param_.learning_rate);
+    } else {
+      perf_monitor.TickStart();
+      this->ApplySplit(nid, gmat, column_matrix, hist_, *p_fmat, p_tree);
+      perf_monitor.UpdatePerfTimer(TreeGrowingPerfMonitor::timer_name::APPLY_SPLIT);
+      int left_id = (*p_tree)[nid].LeftChild();
+      int right_id = (*p_tree)[nid].RightChild();
+      temp_qexpand_depth->push_back(ExpandEntry(left_id,
+                                               p_tree->GetDepth(left_id), 0.0, (*timestamp)++));
```
Comment (hcho3) [inline src/tree/updater_quantile_hist.cc]:
```suggestion
                                                p_tree->GetDepth(left_id), 0.0, (*timestamp)++));
```

File: src/tree/updater_quantile_hist.cc [hunk 38]
```diff
@@ -89,119 +88,286 @@ bool QuantileHistMaker::UpdatePredictionCache(
...
+      temp_qexpand_depth->push_back(ExpandEntry(right_id,
+                                               p_tree->GetDepth(right_id), 0.0, (*timestamp)++));
```
Comment (hcho3) [inline src/tree/updater_quantile_hist.cc]:
```suggestion
                                                p_tree->GetDepth(right_id), 0.0, (*timestamp)++));
```

File: src/tree/updater_quantile_hist.cc [hunk 39]
```diff
@@ -89,120 +88,275 @@ bool QuantileHistMaker::UpdatePredictionCache(
...
+    const GHistIndexBlockMatrix& gmatb,
+    const ColumnMatrix& column_matrix,
+    DMatrix* p_fmat,
+    RegTree* p_tree,
+    const std::vector<GradientPair>& gpair_h) {
+
+  unsigned timestamp = 0;
+  int num_leaves = 0;
+
   for (int nid = 0; nid < p_tree->param.num_roots; ++nid) {
```
Comment (hcho3) [inline src/tree/updater_quantile_hist.cc]: Once this is merged, I will file a follow-up PR to deprecate `num_roots` parameter.

**Comment by CodingCat** [comment]
thanks for the review @trivialfis @hcho3 @RAMitchell

**Comment by Denisevi4** [comment]
So depthwidth is now going to be per-level? :( 

That's unfortunate, because we use depthwidth and the per-nodeness of fast histogram to do feature selection - addding penalty to loss change for unused features. Once feature is used, no penalty is applied. With per-levelness it wouldn't work well especially for deep trees. Was it really worth it?

**Comment by CodingCat** [comment]
@Denisevi4 it helps for reducing communication overhead in distributed training and serves as the base for improving multi-cores performance in the next release by increasing maximum parallelism

**Comment by hcho3** [comment]
@Denisevi4 The per-nodeness of fast histogram was not so good for performance, due to 1) extra syncing needed in distributed setting, and 2) lack of parallelism in multi-core CPU
